Add Base Docker for Development Environment

KazWolfe opened the pull request:
This pull request adds a `Dockerfile` and `docker-compose.yml` configuration for the bot. It builds against Python 3.9 for both dev-like and prod-like environments.

Of note in this Docker configuration:

* Bind mounds are not provided by default through `docker-compose.yml`. This can be overridden by end developers if they so choose, but exists to provide easier compatibility and reproducibility in IDE-based development situations.
* Poetry's `venv` components are disabled entirely, instead installing things to system directly. As this is happening within a Docker context, this should be fine.
* Redis is ready at `redis://redis:6379` with persistence enabled and stored to a volume. By default, Redis is also exposed on host port 6379.
* An environment variable `APP_ENV` exists, which when set to `production` will disable the installation of dev-level dependencies as specified in `pyproject.toml`.

KazWolfe committed:
Add Docker Configs for Dev Environment

- Add a Dockerfile for image build
  - Based on Python 3.9
  - Disables Poetry venv
  - Allows override for prod mode
  - Calls bot directly, better compat with IDEs.
- Add docker-compose.yml sample for development
  - Hosts a Redis server on latest version
  - Reads .env directly
  - DOES NOT contain bindmounts for app folder
- Add entries to .gitignore
  - Add JetBrains config files
  - Add Docker-related dev overrides

#qol #dev-env #ci #docker #no-code

jb3 commented on Dockerfile:
```diff
@@ -0,0 +1,22 @@
+FROM python:3.9
```
Could we use the `-slim` image here?
```suggestion
FROM python:3.9-slim
```

jb3 commented on docker-compose.yml:
```diff
@@ -0,0 +1,20 @@
+    env_file: .env
+    environment:
+      REDIS_URL: "redis://redis:6379/1"
+    links:
+      - redis
```
Is links required here? By default the containers will be able to access others, [links is for defining aliases AFAICS](https://docs.docker.com/compose/networking/#links).

jb3 commented on docker-compose.yml:
```diff
@@ -0,0 +1,20 @@
+version: "3.9"  # optional since v1.27.0
+services:
+  bot:
```
We should use a [`depends_on`](https://docs.docker.com/compose/compose-file/compose-file-v3/#depends_on) option for the bot to ensure Redis spins up before the bot starts.

KazWolfe replied:
`links`  already implies `depends_on`, but if we're removing `links` then yes, this would make sense.

> Links also express dependency between services in the same way as depends_on, so they determine the order of service startup.

jb3 reviewed (changes requested):
Cheers for the contribution, left a few minor suggestions.

KazWolfe committed:
FIX: Update Docker base image to Python 3.9-slim

- Update the docker base image to python:3.9-slim

Required per CR comments.

#cr-fix #qol #dev-env #ci #docker #no-code

KazWolfe committed:
FIX: Remove deprecated "links" in docker-compose

- Replaces "links" directive with "depends_on" directive, as the former is deprecated and should be replaced with either user networks (implicitly used here) and depends_on.

Required per CR comments.

#cr-fix #qol #dev-env #ci #docker #no-code
